### PR TITLE
Add tests for AuthenticatedDAO

### PR DESCRIPTION
**Pull request by TW80000**
* Adds tests for AuthenticatedDAO
* Adds a `TestUtils` class with testing helper methods
* Adds a new context that throws errors when it's used. The point of creating this and using it in AuthenticatedDAO is to prevent users from using the non-context-oriented methods, which they shouldn't be allowed to use for a DAO decorated by AuthenticatedDAO. This way, if a user tries to use a non-context-oriented method, an exception with a helpful message will be thrown.

### CODE REVIEW

**Comment by TW80000** [comment]
@kgrgreer Please review.

File: src/foam/core/InvalidX.java [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+  public X putFactory(Object key, XFactory factory) {
+    throw getException();
+  }
+
+  @Override
+  public String toString() {
+    throw getException();
+  }
+
+  private AccessControlException getException() {
```
Comment (kgrgreer) [inline src/foam/core/InvalidX.java]: I would make the message an argument which would make this class more general/useful.

File: src/foam/dao/AuthenticatedDAO.java [hunk 2]
```diff
@@ -1,17 +1,20 @@
 /**
  * @license
- * Copyright 2017 The FOAM Authors. All Rights Reserved.
+ * Copyright 2018 The FOAM Authors. All Rights Reserved.
```
Comment (kgrgreer) [inline src/foam/dao/AuthenticatedDAO.java]: According to Google lawyers, you just leave the copyright as the year the file was first created.

File: src/foam/test/TestUtils.java [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+public class TestUtils {
+
+  /**
+   * Mock out a DAO in the given context by replacing it with an undecorated, empty MDAO.
+   * Will keep the same 'of' type.
+   * @param x The context you're replacing the DAO in.
+   * @param daoName The name of the DAO in the context you'd like to mock out.
+   * @return A subcontext with the given DAO replaced with an MDAO of the same type.
+   */
+  public static X mockDAO(X x, String daoName) {
```
Comment (kgrgreer) [inline src/foam/test/TestUtils.java:35]: Interesting.

File: src/foam/test/TestUtils.java [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+    user.setLastName("Smith");
+    user.setBusinessName("Widgets International");
+    user.setEmail("john@example.com");
+    File profilePicFile = new File();
+    profilePicFile.setFilename("Profile picture");
+    user.setProfilePicture(profilePicFile);
+    Address businessAddress = new Address();
+    businessAddress.setCity("Toronto");
+    user.setBusinessAddress(businessAddress);
+    Phone businessPhone = new Phone();
```
Comment (kgrgreer) [inline src/foam/test/TestUtils.java]: Business Phone (and possibly other properties as well) isn't part of the standard User model, it's refined in in net.nanopay.auth.Model. This means that this code wouldn't compile for anyone other than nanopay.

**Comment by TW80000** [comment]
Comments addressed. Please review again.

**Review by kgrgreer — APPROVED** [review]

**Pull request merged**